test/e2e/framework.go

```diff
@@ -2279,3 +2281,107 @@ func (data *TestData) waitForDaemonSetPods(timeout time.Duration, dsName string,
 	}
 	return nil
 }
+
+func (data *TestData) createStatefulSet(name string, ns string, size int32, ctrName string, image string, cmd []string, args []string, mutateFunc func(*appsv1.StatefulSet)) (*appsv1.StatefulSet, func() error, error) {
+	podSpec := corev1.PodSpec{
+		Tolerations: []corev1.Toleration{
+			controlPlaneNoScheduleToleration(),
+		},
+		Containers: []corev1.Container{
+			{
+				Name:            ctrName,
+				Image:           image,
+				ImagePullPolicy: corev1.PullIfNotPresent,
+				Command:         cmd,
+				Args:            args,
+			},
+		},
+	}
+	stsSpec := appsv1.StatefulSetSpec{
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"antrea-e2e": name,
+			},
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+				Labels: map[string]string{
+					"antrea-e2e": name,
+				},
+			},
+			Spec: podSpec,
+		},
+		UpdateStrategy:       appsv1.StatefulSetUpdateStrategy{},
+		Replicas:             &size,
+		RevisionHistoryLimit: nil,
+	}
+	sts := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				"antrea-e2e": name,
+			},
+		},
+		Spec: stsSpec,
+	}
+	mutateFunc(sts)
+	resSTS, err := data.clientset.AppsV1().StatefulSets(ns).Create(context.TODO(), sts, metav1.CreateOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() error {
+		if err := data.clientset.AppsV1().StatefulSets(ns).Delete(context.TODO(), name, metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return resSTS, cleanup, nil
+}
+
+func (data *TestData) updateStatefulSetSize(name string, ns string, size int32) (*appsv1.StatefulSet, error) {
+	sts, err := data.clientset.AppsV1().StatefulSets(ns).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	sts.Spec.Replicas = &size
+	resSTS, err := data.clientset.AppsV1().StatefulSets(ns).Update(context.TODO(), sts, metav1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return resSTS, nil
+}
+
+func (data *TestData) restartStatefulSet(name string, ns string) (*appsv1.StatefulSet, error) {
+	sts, err := data.clientset.AppsV1().StatefulSets(ns).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if sts.Annotations == nil {
+		sts.Annotations = map[string]string{}
+	}
+	sts.Annotations[statefulSetRestartAnnotationKey] = time.Now().UTC().Format(time.RFC3339)
```

Comment:
Could you comment why this is needed?

Reply:
Added comment.
`// Modify StatefulSet annotation to trigger a restart for StatefulSet Pods.`

Reply:
Is this an annotation with special meaning? I don't get why annotating the statefulSet will trigger restart of its Pods. Do you mean the annotation of the podTemplate?

Reply:
Yes, this is a StatefulSet annotation, which can trigger a StatefulSet restart if its value changed.
It is equivalent to `kubectl rollout restart xxx`
https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/objectrestarter.go

Reply:
The code in the link updates the annotation of the podTemplate, not the StatefulSet. I think it makes sense as updating any field of the podTemplate will lead to Pod recreation. But this is not the case of updating annotation of StatefulSet.
I think the current `restartStatefulSet` won't really trigger Pod recreation, have you confirmed by watching the Pod events?

Reply:
Since "kubectl.kubernetes.io/restartedAt" doesn't have special meaning (the prefix just means it's triggerred by kubectl), it doesn't make sense to use it in antrea e2e, which could be confusing. We could use a more meaningful annotation `antrea-e2e/restartedAt: time.Now().UTC().Format(time.RFC3339)`

Reply:
Corrected the annotation as it is a PodTemplate annotation.
Changed annotation key to `antrea-e2e/restartedAt`